### ISSUE TITLE
vendoring a project that itself uses godep is broken

### ISSUE THREAD

**Issue by thockin** [tools/godep#224]
I am trying to vendor all of github.com/coreos/etcd, which uses godep with rewriting.  We do not use rewriting.

I ran `go get` and it works.

I ran `godep save` and it complains that a handful of deps are not found.

Nobody references those deps directly - they are only referenced from within etcd by their re-written paths.  godep's unqualify() function strips off the rewritten prefix, even though that prefix is not my project's prefix and I am not using rewriting.  

Is the assumption that I manually figure out which of my deps are also using godep and manually repeat the `godep restore` process for each?

**Comment by freeformz** [comment]
For each of the deps that aren't found, run `go get -u <dep>` to make sure you have them installed locally and then you will be able to run `godep save`.

TBH though I'm not 100% sure that the side effects of using etcd like this.

**Comment by freeformz** [comment]
@kr Should we instead resolve these dependencies to the versions vendored in the sub project?

For instance `github.com/coreos/etcd/raft` imports `github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context`. ATM you have to have `golang.org/x/net/context` in your $GOPATH to be able to run `godep save`.

**Comment by kr** [comment]
The approach I've tried to take so far is to let package main decide which revision of each dependency to use. Normally, a dependency package that itself has further dependencies should not even use godep. A repo like etcd is a little more confusing because it also has some commands, so they vendor all their deps. But when using packages from etcd in another project, we should try to treat the etcd packages as if they hadn't vendored anything.

Godep could try to use etcd's recorded revision of those leaf dependencies, but that strategy doesn't always work: it's possible to have two repos that have each vendored a different revision of a leaf package. Our approach so far is to use a simple rule consistently: always let the user choose. The way the user expresses this choice is by installing their chosen revision of each dependency (leaf or otherwise) into their $GOPATH.

However, there's certainly still room for making this more convenient. If a project has a lot of dependency repos like etcd, it's tedious to run `godep restore` for each one. Godep could potentially make this process more automatic. But note that that if the same leaf dependency has been vendored in multiple repos, the user will be left with an arbitrary revision of it in GOPATH—whichever one was restored last. Maybe that's fine, but I'm a little worried that folks might not be aware of that. It could go unnoticed for a long time, and might be an unwelcome surprise.

**Comment by thockin** [comment]
> when using packages from etcd in another project, we should try to treat
the etcd packages as if they hadn't vendored anything

This is tricky.  If I want to vendor all of etcd, I want to use the deps
that etcd wants.  If I am vendoring an etcd lib that has transitive deps,
it is less obvious.

**Comment by kr** [comment]
It is a basic assumption of godep that you are vendoring packages, not commands.

I agree that it would be useful to vendor commands along with their dependencies, but as you have indicated, that's a different thing, and godep doesn't really do that (yet). It might sort of work, but if so, that is a consequence of godep not being careful enough about what it does, not a deliberate feature.

**Comment by thockin** [comment]
Well with Go's new vendor feature, godep won't have a say in what I do or
don't sub-vendor.

**Comment by apriendeau** [comment]
@kr It shouldn't break when someone does use their own packages. Recurisiviness will be a problem but that means that you should still work. For example, `godep save` fails with [go-dockerclient](https://github.com/fsouza/go-dockerclient).

**Comment by freeformz** [comment]
It worked for me: https://github.com/fsouza/go-dockerclient/pull/341

**Comment by apriendeau** [comment]
I forgot to mention, I was using the vendor experiment and 1.5beta3

**Comment by apriendeau** [comment]
It looks like you are replacing party with godeps.  I was referring to code that depends on go-dockerclient as it is now

**Comment by freeformz** [comment]
@apriendeau Do you have an example repo and steps to reproduce?

**Comment by apriendeau** [comment]
@freeformz if you use the first example as new project then run:
```shell
export GO15VENDOREXPERIMENT=1
go get ./...
package github.com/apriendeau/recreate-godeps
        imports github.com/fsouza/go-dockerclient
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/archive
        imports github.com/fsouza/go-dockerclient/vendor/github.com/Sirupsen/logrus: must be imported as github.com/Sirupsen/logrus
package github.com/apriendeau/recreate-godeps
        imports github.com/fsouza/go-dockerclient
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/archive: must be imported as github.com/docker/docker/pkg/archive
package github.com/apriendeau/recreate-godeps
        imports github.com/fsouza/go-dockerclient
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/fileutils: must be imported as github.com/docker/docker/pkg/fileutils
package github.com/apriendeau/recreate-godeps
        imports github.com/fsouza/go-dockerclient
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/archive
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/pools
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/ioutils: must be imported as github.com/docker/docker/pkg/ioutils
package github.com/apriendeau/recreate-godeps
        imports github.com/fsouza/go-dockerclient
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/archive
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/pools: must be imported as github.com/docker/docker/pkg/pools
package github.com/apriendeau/recreate-godeps
        imports github.com/fsouza/go-dockerclient
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/archive
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/promise: must be imported as github.com/docker/docker/pkg/promise
package github.com/apriendeau/recreate-godeps
        imports github.com/fsouza/go-dockerclient
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/stdcopy: must be imported as github.com/docker/docker/pkg/stdcopy
package github.com/apriendeau/recreate-godeps
        imports github.com/fsouza/go-dockerclient
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/archive
        imports github.com/fsouza/go-dockerclient/vendor/github.com/docker/docker/pkg/system: must be imported as github.com/docker/docker/pkg/system
godep save
godep: cannot find package "github.com/docker/docker/pkg/archive" in any of:
        /usr/local/opt/go/libexec/src/github.com/docker/docker/pkg/archive (from $GOROOT)
        /Users/austin/Go/src/github.com/docker/docker/pkg/archive (from $GOPATH)
godep: cannot find package "github.com/docker/docker/pkg/fileutils" in any of:
        /usr/local/opt/go/libexec/src/github.com/docker/docker/pkg/fileutils (from $GOROOT)
        /Users/austin/Go/src/github.com/docker/docker/pkg/fileutils (from $GOPATH)
godep: cannot find package "github.com/docker/docker/pkg/ioutils" in any of:
        /usr/local/opt/go/libexec/src/github.com/docker/docker/pkg/ioutils (from $GOROOT)
        /Users/austin/Go/src/github.com/docker/docker/pkg/ioutils (from $GOPATH)
godep: cannot find package "github.com/docker/docker/pkg/pools" in any of:
        /usr/local/opt/go/libexec/src/github.com/docker/docker/pkg/pools (from $GOROOT)
        /Users/austin/Go/src/github.com/docker/docker/pkg/pools (from $GOPATH)
godep: cannot find package "github.com/docker/docker/pkg/promise" in any of:
        /usr/local/opt/go/libexec/src/github.com/docker/docker/pkg/promise (from $GOROOT)
        /Users/austin/Go/src/github.com/docker/docker/pkg/promise (from $GOPATH)
godep: cannot find package "github.com/docker/docker/pkg/stdcopy" in any of:
        /usr/local/opt/go/libexec/src/github.com/docker/docker/pkg/stdcopy (from $GOROOT)
        /Users/austin/Go/src/github.com/docker/docker/pkg/stdcopy (from $GOPATH)
godep: cannot find package "github.com/docker/docker/pkg/system" in any of:
        /usr/local/opt/go/libexec/src/github.com/docker/docker/pkg/system (from $GOROOT)
        /Users/austin/Go/src/github.com/docker/docker/pkg/system (from $GOPATH)
godep: error loading dependencies
```

**Comment by freeformz** [comment]
@apriendeau Sorry, I am lost. I am not sure what your starting point is above. 

Can you provide a step by step repro from a clean $GOPATH?

**Comment by peter-edge** [comment]
@apriendeau @freeformz of note, you should pull on go-dockerclient, I preemptively fixed the go 1.5 vendoring issue with https://github.com/fsouza/go-dockerclient/pull/339

**Comment by timothysc** [comment]
@thockin I'm hitting the same issue with trying to pull in a sub-element of etcd. 
/cc @yichengq

**Comment by freeformz** [comment]
@timothysc Can you provide more information about what problem you are experiencing and/or put together a publicly accessible reproduction?

**Comment by newhook** [comment]
He's running into this:

godep: cannot find package "golang.org/x/net/context" in any of:
	/Users/matthew/.gvm/gos/go1.4.2/src/golang.org/x/net/context (from $GOROOT)
	/Users/matthew/go/src/golang.org/x/net/context (from $GOPATH)
	/Users/matthew/src/services/src/golang.org/x/net/context
	/Users/matthew/src/services/.gopack/vendor/src/golang.org/x/net/context
godep: error loading dependencies

etcd vendors this as github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context and furthemore documents in its example that this should be imported as such.

**Comment by freeformz** [comment]
ATM, if you are vendoring etcd and it uses Godep to vendor net/context and you want to use the version of net/context that etcd vendors then you need to import it as:
`your.package/Godeps/_workspace/src/github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context`

Given the go vendor experiment becoming the standard in 1.6 I'm not sure it's worth putting much effort behind `godep save` being able to handle this case.

Thoughts?

**Comment by thockin** [comment]
It's effectively impossible to vendor a project like etcd that uses godep rewriting with GO15VENDOREXPERIMENT.  As an example, we vendor from rkt, which does uses rewriting.  Error:

```
... cannot use apisvcConn (type *"google.golang.org/grpc".ClientConn) as type *"github.com/coreos/rkt/Godeps/_workspace/src/google.golang.org/grpc".ClientConn ...
```

rkt is expecting a `ClientConn` from github.com/coreos/rkt/Godeps/_workspace/src/google.golang.org/grpc, but my code is passing it from google.golang.org/grpc.  You can argue that the rkt API should not be exposing any types from other pkgs, but it is.

How is one supposed to get past this, given that there are and will be repos that use Godep in the old way for a while, yet.  Can we teach godep to recognize this and un-rewrite?

**Comment by freeformz** [comment]
FWIW: This is also an issue with any nested vendor directory setup. I opened (or commented anyway) on a similar issue against Go itself. The only real way to handle this is to flatten the deps so there is only one  `google.golang.org/grpc` package in use. I'd like to teach godep to do this (it already does it for the older workspace format). It needs to be taught how to check *both* and DTRT. Hopefully later this month unless someone else shows up with PRs.

**Comment by thockin** [comment]
It would be nice to have a choice - some libs are ONLY consumed by a dep
internally, and having 2 versions linked is not so bad.  Other times you
need to flatten, like this grpc case.

**Comment by freeformz** [comment]
@thockin While it may be nice, I wonder if the possible vectors for confusion would be worth it.

**Comment by thockin** [comment]
I agree it might be confusing, and would need REALLY good tooling to make
it clear how things are currently configured and to allow
management/changes.  OTOH, library L which uses dep D and exposes an API to
app A should not expose D's type.  It would be much better to take and
return L-defined interfaces.  But that's a different argument.

Note that this represents a reversal from me - I was initially against
flattening.

**Comment by dlorenc** [comment]
@thockin How did you get past this? I'm now hotting the same thing in minikube trying to update to the latest kubernetes build.

**Comment by thockin** [comment]
Are you sure this is what you are hitting? Converting our main repo mostly
"just worked", including recursive godeps.

Start with a clean GOPATH
    godep restore
    rm -rf Godeps
    godep save ./...
    git add vendor
    git add -f $(git ls-files --other vendor/)

Look for other files under Godeps that you might want to restore (like the
license state)

**Comment by dlorenc** [comment]
Yeah it was the same error, but I just got past it somehow. I blew away my Godeps folder, re-ran "godep save" and then it worked the second time...

**Comment by freeformz** [comment]
Closing issues before archiving. Thank You\!
